### DISCUSSION TITLE
Unchained as B2C Multi-Vendor system?

### DISCUSSION THREAD

**Discussion by pozylon** [unchainedshop/unchained#381] [Q&A] [answered]
Originally asked by @uyasarkocal. https://github.com/unchainedshop/unchained/issues/364

> "Any plans on making it into a multivendor system? Or if I decide to do it, would you recommend it architecture wise?"

We know of an agency that wants to build a multi-tenant system and we have received various requests about using Unchained in a multi-vendor environment.

Multi-tenant: Hosting multiple shops for different retailers with their own userbase, catalog on one unchained engine instance
Multi-vendor / Multi-sided platform: Hosting a single unchained instance that allows vendors to login and submit articles for sale

At the moment, Unchained does not support either of those cases out of the box, but the ACL system is very powerful and could be extended with some effort to support both cases.

For multi-tenant, you will have to lock down all queries and mutations until the http request contains either a cookie or an http header setting the tenant id like X-Unchained-Tenant (easy), then based on the tenant id select different mongodb database. This requires a complete abstraction of the current db access layer as a base (already planned, see here https://github.com/unchainedshop/unchained/issues/115, we will remove meteor's mongodb and replace it with the native driver)

For multi-vendor, you will have to extend the ACL with more roles (clientManager, productManager) write very project specific ACL rules, add custom resolvers for the products you want to have created and write unchained plugins to scope visibility of drafts to certain users. As important as that, you will have to create an own Admin UI leveraging all of that which is maybe even more work than extending the backend. We are very interested in helping anybody out achieving this with our e-commerce engine and we are very sure it's the right technology.

So in short:

Achieving multi-tenancy: High Complexity, requires fundamental changes to the core first but basically no frontend changes required
Achieving multi-vendor: Low Complexity, requires a lot of additional code and project specific development and a custom admin frontend to support the case (you cannot use controlpanel)

**Comment by pozylon** [comment, marked as answer]
Answer inline
